File: src/main/java/org/snipsniper/configwindow/StampJPanel.java
package org.snipsniper.configwindow;

import org.snipsniper.ImageManager;
import org.snipsniper.sceditor.stamps.IStamp;
import org.snipsniper.utils.Utils;
import org.snipsniper.utils.Vector2Int;

import javax.swing.*;
import java.awt.*;
import java.awt.image.BufferedImage;

public class StampJPanel extends JPanel {
    private IStamp stamp;
    private BufferedImage background;
    private boolean backgroundEnabled = true;
    private RenderingHints qualityHints;

    private int margin = 0;

    public void setStamp(IStamp stamp) {
        this.stamp = stamp;
        repaint();
    }

    public void setMargin(int margin) {
        this.margin = margin;
    }

    public void setBackground(BufferedImage image) {
        background = image;
    }

    public void setBackgroundEnabled(boolean enabled) {
        backgroundEnabled = enabled;
        repaint();
    }

    public void paint(Graphics g) {
        super.paint(g);

        int sizeX = getWidth() - margin;
        int sizeY = getHeight() - margin;

        Graphics2D g2d = (Graphics2D) g;

        g.setColor(getBackground());
        g.fillRect(0, 0, getWidth(), getHeight());

        if(qualityHints == null) {
            qualityHints = Utils.getRenderingHints();
        }

        g2d.setRenderingHints(qualityHints);

        if(background != null && backgroundEnabled) {
            if(!(background.getWidth() >= getWidth() && background.getHeight() >= getHeight()))
                background = ImageManager.getCodePreview();
            int pos = margin / 2;
            int width = sizeX + margin / 2;
            int height = sizeY + margin / 2;
            g2d.drawImage(background, pos, pos, width, height, pos, pos, width, height, null);
            Color oldColor = g2d.getColor();
            g2d.setColor(Color.BLACK);
            g2d.drawRect(pos, pos, width - margin / 2, height - margin / 2);
            g2d.setColor(oldColor);
        }
        if(!isEnabled())
            g2d.setColor(Utils.getDisabledColor());
        g2d.drawRect(0,0,getWidth() - 1,getHeight() -1);

        if(stamp != null)
            stamp.render(g2d, null, new Vector2Int(getWidth()/2, getHeight()/2), new Double[]{1D, 1D}, false, false, 0);

        if(!isEnabled())
            g2d.fillRect(0, 0, getWidth(), getWidth()/2);
        g2d.dispose();
    }

}

File: src/main/kotlin/net/snipsniper/configwindow/StampJPanel.kt
package net.snipsniper.configwindow

import net.snipsniper.ImageManager
import net.snipsniper.sceditor.stamps.IStamp
import net.snipsniper.utils.Utils
import org.capturecoop.ccutils.math.CCVector2Int
import java.awt.Color
import java.awt.Graphics
import java.awt.Graphics2D
import java.awt.image.BufferedImage
import javax.swing.JPanel

class StampJPanel(stamp: IStamp, private var background: BufferedImage? = null, private var margin: Int = 0): JPanel() {
    private val renderingHints = Utils.getRenderingHints()
    var backgroundEnabled = true
        set(value) {
            field = value
            repaint()
        }

    var stamp = stamp
        set(value) {
            field = value
            repaint()
        }

    override fun paint(g: Graphics) {
        super.paint(g)
        g.color = getBackground()
        g.fillRect(0, 0, width, height)

        val g2d = g as Graphics2D
        g2d.setRenderingHints(renderingHints)
        if(backgroundEnabled) {
            background?.let { bg ->
                if(!(bg.width >= width && bg.height >= height))
                    background = ImageManager.getCodePreview()
                val pos = margin / 2
                val width = width - margin + margin / 2
                val height = height - margin + margin / 2
                g2d.drawImage(background, pos, pos, width, height, pos, pos, width, height, null)
                g2d.color.also { oldColor ->
                    g2d.color = Color.BLACK
                    g2d.drawRect(pos, pos, width - margin / 2, height - margin / 2)
                    g2d.color = oldColor
                }
            }
        }
        if(!isEnabled) g2d.color = Utils.getDisabledColor()
        g2d.drawRect(0,0, width - 1, height -1)

        stamp.render(g2d, null, CCVector2Int(width / 2, height / 2), doubleArrayOf(1.0, 1.0).toTypedArray(), false, false, 0)

        if(!isEnabled) g2d.fillRect(0, 0, width, width / 2)
        g2d.dispose()
    }
}